Updated JITHelpers class to avoid false positives

qasimy123 opened the pull request:
Created methods to allow an expected exception to be set from the test framework.
Verifies if the thrown exception is unexpected before running
the debugging agent. More details are found in #5.

The JITHelpers class now supports:
- Setting an expected exception
- Getting an expected exception
- Weed out false positives from triggering debugging agent.

Closes: #5
Signed-off-by: Qasim Khawaja <khawaja@ualberta.ca>


Here are some further tests that I ran to verify my implementation:
[Multi Threaded (expected exception thrown with no bug enabled)](https://github.com/r30shah/openj9-jit-debug-agent/files/7524599/multi-threaded-expected-exception-no-bug.log)
[Multi Threaded (unexpected exception thrown with no bug enabled](https://github.com/r30shah/openj9-jit-debug-agent/files/7524600/multi-threaded-unexpected-exception-no-bug.log)
[Single Threaded (bug enabled)](https://github.com/r30shah/openj9-jit-debug-agent/files/7524602/single-threaded-bug-enabled.log)

### Callout
@r30shah These are two more tests that I ran with the bug enabled with the multithreaded tests. No synchronization causes the debug agent to end with a trace log and the same exception trace from #7. ( I have your latest fix)
I was able to resolve this by turning the block that runs the agent to be synchronized. (Line 1246)
[not-synced-multithreaded-bug-enabled.log](https://github.com/r30shah/openj9-jit-debug-agent/files/7524601/not-synced-multithreaded-bug-enabled.log)
[synced-multithreaded-bug-enabled.log](https://github.com/r30shah/openj9-jit-debug-agent/files/7524603/synced-multithreaded-bug-enabled.log)

r30shah commented:
@qasimy123  Can rebase your branch? Seems like all the commits we have for JIT debug agent is included in this PR.

qasimy123 commented:
@r30shah The conflicts should be resolved.

qasimy123 commented:
@r30shah I got the debug agent to trigger with the bug enabled.

Here is the output:
https://github.com/r30shah/openj9-jit-debug-agent/files/7524603/synced-multithreaded-bug-enabled.log

With the following command
```console
/openj9-openjdk-jdk11/build/linux-x86_64-normal-server-release/jdk/bin/java -Xdump:none -Xnoaot -Xcheck:jni -Xjit:forceUsePreexistence -cp ./junit-4.12.modified.jar:./hamcrest-all-1.3.jar: DebugAgentRunner 100000 &> synced-multithreaded-bug-enabled.log
```

This is the output I get running the command you provided. I noticed that it doesn't exit cleanly and theres a System dump generated.
[rahil.log](https://github.com/r30shah/openj9-jit-debug-agent/files/7558353/rahil.log)

```console
/openj9-openjdk-jdk11/build/linux-x86_64-normal-server-release/jdk/bin/java -Xshareclasses:none -Xnoaot -Xcheck:jni -Xjit:forceUsePreexistence,"{Test_String.*13*}(traceFull,log=methodLog)"  --add-exports java.base/com.ibm.jit=ALL-UNNAMED -cp junit-4.12.modified.jar:hamcrest-all-1.3.jar:. DebugAgentRunner 640000 &> rahil.log
```

fjeremic commented on jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java:
```diff
@@ -69,6 +70,49 @@ private static JITHelpers jitHelpers() {
+
+	public static void initExceptionMap() {
+		if (exceptionMap == null) {
+			exceptionMap = new java.util.HashMap<String, Class<? extends Throwable>>();
```
We should probably synchronize this method on `JitHelpers.class` since multiple threads can be initializing the map.

qasimy123 replied:
@r30shah @fjeremic 
I believe this may have been causing the issue with the wrong method being picked up. When I tested it with synchronization it went away. Ref: https://github.com/r30shah/openj9-jit-debug-agent/issues/5#issuecomment-973206780
[non-working-example.log](https://github.com/r30shah/openj9-jit-debug-agent/files/7573901/non-working-example.log)
[working-example.log](https://github.com/r30shah/openj9-jit-debug-agent/files/7573902/working-example.log)

r30shah commented on jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java:
```diff
@@ -69,6 +70,49 @@ private static JITHelpers jitHelpers() {
+		initExceptionMap();
+		
+		if (expectedException == null) {
+			System.out.println("Removing exception for thread: " + vmThreadName);
```
We do not need to print the message everytime it adds or removes the exception in the map. This would affect the timing as well.

r30shah commented:
@qasimy123  Can you squash the commits to single one. I am OK if you squash my commits that adds debug prints and removes some print message with your single commit.

r30shah commented on jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java:
```diff
@@ -69,6 +70,47 @@ private static JITHelpers jitHelpers() {
 		return helpers;
 	}
 
+	/*
```
Can you add doxygen format comments? You can checkout the references of such comments on Various JCL code for example https://github.com/eclipse-openj9/openj9/blob/73dac1a16e721e8d62bf2ac6bec116e4ec4464c4/jcl/src/java.base/share/classes/java/lang/Class.java#L345-L353

r30shah commented on jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java:
```diff
@@ -69,6 +70,47 @@ private static JITHelpers jitHelpers() {
+		}
+	}
+
+	public static void destroyExceptionMap() {
```
I do not see you using this function at all. In which cases do you expect this method to be called?

qasimy123 replied:
I think this can be removed, I can't think of a scenario when this would need. I'll update that.

r30shah reviewed (changes requested):
@qasimy123  Overall changes looks Good to me, Some minor questions I had asked in the the review. Also for the methods you are adding to this method for this project (I would include the `invoke` method as well, please add comments.

r30shah commented:
@qasimy123   You need to reword the commit message. You can remove everything after `signoff` line. We do not need any commit message for the changes you committed as part of the review response.

r30shah commented on jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java:
```diff
@@ -69,6 +70,50 @@ private static JITHelpers jitHelpers() {
 	}
 
+	/**
+	 * Public method to set the expected exception for the current thread.
```
No need to mention Public method, method signature shows that information.

r30shah commented on jcl/src/java.base/share/classes/com/ibm/jit/JITHelpers.java:
```diff
@@ -69,6 +70,50 @@ private static JITHelpers jitHelpers() {
+	}
+
+	/**
+	 * Public method to get the expected exception for the current thread.
```
Same as above.

r30shah reviewed (approved):
@qasimy123  Minor nitpicks, Overall change looks good to me

qasimy123 committed:
Updated JITHelpers class to avoid false positives

Created methods to allow an expected exception to be set from test framework.
Verifies if the thrown exception is unexpected before running
the debugging agent. More details found in #5.

The JITHelpers class now supports:
- Setting an expected exception
- Getting an expected exception
- Weeds out false positives from triggering debugging agent.

Closes: #5
Signed-off-by: Qasim Khawaja <khawaja@ualberta.ca>

r30shah commented:
Changes looks good to me, @fjeremic  wanted to check with you before merging this.

r30shah commented:
Merging this one as PR has been opened for Debug Agent using ASM against openj9-util
